airbyte-integrations/connectors/source-bing-ads/source_bing_ads/reports.py

```diff
@@ -0,0 +1,252 @@
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+#
+
+from abc import ABC, abstractmethod
+from datetime import datetime
+from typing import Any, Iterable, Mapping, MutableMapping, Optional, Union
+
+import pendulum
+import source_bing_ads.source
+from airbyte_cdk.models import SyncMode
+from bingads.v13.internal.reporting.row_report import _RowReport
+from bingads.v13.internal.reporting.row_report_iterator import _RowReportRecord
+from bingads.v13.reporting import ReportingDownloadParameters
+from suds import sudsobject
+
+REPORT_FIELD_TYPES = {
+    "AccountId": "integer",
+    "AdId": "integer",
+    "AdGroupCriterionId": "integer",
+    "AdGroupId": "integer",
+    "AdRelevance": "number",
+    "Assists": "integer",
+    "AverageCpc": "number",
+    "AveragePosition": "number",
+    "BusinessCategoryId": "integer",
+    "BusinessListingId": "integer",
+    "CampaignId": "integer",
+    "ClickCalls": "integer",
+    "Clicks": "integer",
+    "ConversionRate": "number",
+    "Conversions": "number",
+    "CostPerAssist": "number",
+    "CostPerConversion": "number",
+    "Ctr": "number",
+    "CurrentMaxCpc": "number",
+    "EstimatedClickPercent": "number",
+    "EstimatedClicks": "integer",
+    "EstimatedConversionRate": "number",
+    "EstimatedConversions": "integer",
+    "EstimatedCtr": "number",
+    "EstimatedImpressionPercent": "number",
+    "EstimatedImpressions": "integer",
+    "ExactMatchImpressionSharePercent": "number",
+    "HistoricAdRelevance": "number",
+    "Impressions": "integer",
+    "ImpressionSharePercent": "number",
+    "KeywordId": "integer",
+    "LandingPageExperience": "number",
+    "LowQualityClicks": "integer",
+    "LowQualityClicksPercent": "number",
+    "LowQualityImpressions": "integer",
+    "LowQualitySophisticatedClicks": "integer",
+    "PhoneCalls": "integer",
+    "PhoneImpressions": "integer",
+    "QualityImpact": "number",
+    "QualityScore": "number",
+    "ReturnOnAdSpend": "number",
+    "Revenue": "number",
+    "RevenuePerAssist": "number",
+    "RevenuePerConversion": "number",
+    "SidebarBid": "number",
+    "Spend": "number",
+    "MonthlyBudget": "number",
+    "DailySpend": "number",
+    "MonthToDateSpend": "number",
+}
+
+
+class ReportsMixin(ABC):
+    # The directory where the file with report will be downloaded.
+    file_directory: str = "/tmp"
+    # timeout for reporting download operations in milliseconds
+    timeout: int = 300000
+    report_file_format: str = "Csv"
+    aggregation_disabled: bool = False
+
+    @property
+    @abstractmethod
+    def report_name(self) -> str:
+        """
+        Specifies bing ads report naming
+        """
+        pass
+
+    @property
+    @abstractmethod
+    def report_columns(self) -> Iterable[str]:
+        """
+        Specifies bing ads report naming
+        """
+        pass
+
+    def get_request_date(self, reporting_service, date: datetime) -> sudsobject.Object:
```

Comment:
docstring for this would be useful (why do we need to use reporting_service.factory to create a date?)